README.md

# memer <img src="man/figures/logo.png" align="right" width="120" />

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/sctyner/memer.svg?branch=master)](https://travis-ci.org/sctyner/memer)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/sctyner/memer?branch=master&svg=true)](https://ci.appveyor.com/project/sctyner/memer)
[![CRAN
status](https://www.r-pkg.org/badges/version/memer)](https://cran.r-project.org/package=memer)
<!-- badges: end -->

## Overview

memer is a a `tidyverse`-compatible `R` package for creating memes.
Mostly wrappers around `magick` functions.

## Installation

``` r
# Install the development version from GitHub:
# install.packages("devtools")
devtools::install_github("sctyner/memer")
```

## Available Memes

``` r
library(memer)
meme_list()
```

    ##  [1] "AllTheThings"       "AmericanChopper"    "AncientAliens"     
    ##  [4] "BatmanRobin"        "DistractedBf"       "EvilKermit"        
    ##  [7] "ExpandingBrain"     "FirstWorldProbs"    "FryNotSure"        
    ## [10] "IsThisAPigeon"      "NoneOfMyBusiness"   "CheersLeo"         
    ## [13] "OneDoesNotSimply"   "DosEquisMan"        "OffRamp"           
    ## [16] "OprahGiveaway"      "Philosoraptor"      "PicardFacePalm"    
    ## [19] "PicardWTH"          "PutItPatrick"       "Spongebob"         
    ## [22] "SuccessKid"         "ThatWouldBeGreat"   "TheRockDriving"    
    ## [25] "ThinkAboutIt"       "TrumpBillSigning"   "TwoButtonsAnxiety" 
    ## [28] "WhatIfIToldYou"     "CondescendingWonka" "Y-U-NOguy"         
    ## [31] "ShiaJustDoIt"       "Rainbow"            "Purples"

## Usage

``` r
meme_get("DistractedBf") %>% 
  meme_text_distbf("tidyverse", "new R users", "base R")
```

<img src="README_files/figure-gfm/cdistbf-1.png" width="75%" style="display: block; margin: auto;" />

``` r
meme_get("OprahGiveaway") %>% 
  meme_text_bottom("EVERYONE GETS A %>%!!!", size = 36)
```

<img src="README_files/figure-gfm/oprah-1.png" width="75%" style="display: block; margin: auto;" />

``` r
meme_get("AmericanChopper") %>%
  meme_text_chopper("Stop calling yourself 'just an R user'!",
                    "But I've never written a package!",
                    "So?? You're still an important part of the R community!",
                    "But people who write packages are more important",
                    "NO! All members of the R community are valued & deserve respect!")
```

<img src="README_files/figure-gfm/chopper-1.png" width="40%" style="display: block; margin: auto;" />

## Post R Memes directly to Twitter\!

``` r
meme_get("ExpandingBrain") %>% 
  meme_text_brain("Sharing memes online", 
                  "Making your own memes", 
                  "Making memes in R", 
                  "Using rtweet to share your memes", 
                  size = 17)
```

<img src="README_files/figure-gfm/brain-1.png" width="50%" style="display: block; margin: auto;" />

You can also post memes directly to Twitter. You will need to install
the [rtweet](https://github.com/mkearney/rtweet/) package for that. Just
make sure to set up a Twitter token ([see here for more information on
that](https://rtweet.info/articles/auth.html)) and you are good to go.
The `meme_tweet` function also includes the possibility to tag the [R
Memes for Statistical Fiends](https://twitter.com/rstatsmemes) account
on Twitter for maximum exposure :)

``` r
meme_get("DosEquisMan") %>% 
  meme_text_top("I don't always tweet memes", size = 28) %>% 
  meme_text_bottom("But when I do\nI use the memer package in R", size = 26) %>%
  meme_tweet(tweet_text = "It's dangerous to meme alone. Here take this.",
             tag_rstatsmemes = T)
```

<img src="https://i.imgur.com/XvKhZZc.png" width="75%" style="display: block; margin: auto;" />

\`\`\`


NEWS.md
# memer 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/get-meme.R
#' Read in the meme you wish to create
#'
#'
#' @param memename A character describing the meme to get. See [meme_list()]
#'
#' @examples
#' meme_get("AllTheThings")
#' @export
#' @importFrom magick image_read
#' @importFrom readr read_csv
#' @importFrom utils data
#'
#' @section To add a new meme to the data:
#' \describe{
#'    \item{First}{Add the meme to the inst/extdata folder (png)}
#'    \item{Then}{Add row to the blankmemes data `dplyr::add_row(blankmemes, filename = "...", name = "...")`}
#'    \item{Finally}{Run `usethis::use_data(blankmemes, overwrite = T)`}
#' }
#'
meme_get <- function(memename) {
  if (!is.character(memename)) {
    stop("Error: memename must be a character. See meme_list().")
  }

  idx <- which(memer::blankmemes$name == memename)
  filepath <- paste0("extdata/", memer::blankmemes$filename[idx])

  p <- image_read(system.file(filepath, package = "memer", lib.loc = NULL, mustWork = T))
  return(p)
}

#' List available memes
#'
#'
#' @examples
#' meme_list()
#' @export
meme_list <- function() {
  return(memer::blankmemes$name)
}


R/post-meme.R
#' Posts a meme to Twitter
#'
#' Pipe your meme directly into this function to post your meme on Twitter. Make sure to have `rtweet` installed!
#'
#' @param meme an external pointer of class 'magick-image'
#' @param tweet_text The text that you want to post along with your meme
#' @param tag_rstatsmemes adds two hashtags and tags the "rstatsmemes" account
#' @param token Every user should have their own Oauth (Twitter API) token. By
#'   default `token = rtweet::get_token()`, which looks for a saved
#'   Twitter token via environment variables (which is what [rtweet::create_token()]
#'   sets up by default during initial token creation). For instruction on how
#'   to create a Twitter token see the tokens vignette, i.e.,
#'   `vignettes("auth", "rtweet")` or see `?tokens`. See [rtweet::get_token()].
#'
#' @examples
#'\dontrun{
#' meme_get("OprahGiveaway") %>%
#'   meme_text_bottom("EVERYONE GETS R MEMES!!!", size = 36) %>%
#'   meme_tweet(tweet_text = "It's dangerous to meme alone. Here take this.",
#'   tag_rstatsmemes = T)
#'}
#'
#' @importFrom magick image_write
#' @importFrom stringr str_glue
#'
#' @section How to share your meme on Twitter:
#' \describe{
#'    \item{First}{Authenticate your Twitter account, i.e. create a token)}
#'    \item{Then}{Create meme of your liking}
#'    \item{Finally}{Post meme to Twitter}
#' }
#' @references
#' Kearney, M. W. (2018). rtweet: Collecting Twitter Data. R package version 0.6.7 Retrieved from
#' \url{https://cran.r-project.org/package=rtweet}
#' @export
meme_tweet <- function(meme, tweet_text = "It's dangerous to meme alone. Here take this.",
                       tag_rstatsmemes = FALSE, token = NULL) {
 ## get token
  if (is.null(token)){
    token <- rtweet::get_token()
  }

  ## create temporary filename
  fname <- tempfile(fileext = paste0(".png"))

  ## temporarily write file
  image_write(meme, path = fname)

  ## option to tag rstatsmemes
  if (tag_rstatsmemes == T) {
    tweet_text <- str_glue(tweet_text, "\n\n#rstats #rstatsmemes @rstatsmemes")
  }

  ## post tweet from rtweet
  rtweet::post_tweet(status = tweet_text, media = fname, token = token)

}


R/explain-meme.R
#' Explain a meme (for the old people out there)
#' 
#' Takes a meme name from memer and takes you to the 
#' knowyourmeme page that can explain it for us old
#' people.
#'
#' @param memename A character describing the meme to get. See [meme_list()]
#' @param launch Whether to launch the link in your browser 
#'   (defaults to TRUE if in interactive session)
#' @param content_warning Whether to warn about potentially offensive content
#'   on knowyourmeme
#'
#' @return The link to the knowyourmeme article about a meme
#' @export
#'
#' @examples
#' meme_explain("Hotlinedrake", launch = FALSE)
meme_explain <- function(memename, 
                         launch = interactive(), 
                         content_warning = TRUE) {
  link <- meme_links[meme_links$name == memename, ]
  link <- link$url
  if (length(link) == 0 || is.na(link)) {
    stop(paste0("Could not find link for meme: ", memename), call. = FALSE)
  }
  
  if (launch) {
    if (content_warning) {
      cat("knowyourmeme can contain potentially offensive results,\n  are you sure you want to open it? [Y]es / [N]o:\n")
      result <- tryCatch(scan("", what = "character", quiet = TRUE, nlines = 1), error = function(x) "N")
      if (!identical(toupper(result), "Y")) {
        return(NULL)
      }
    }
    utils::browseURL(link)
    invisible(link)
  } else {
    if (content_warning) {
      message("knowyourmeme can contain potentially offensive results.")
    }
    link
  }
}

R/text-annotate.R
#' Place text in the meme
#'
#'
#' @param img An image of class `magick_image`. See [meme_get()].
#' @param txt A character string
#' @param font string with font family such as `"sans"`, `"mono"`, `"serif"`, `"Times"`, `"Helvetica"`, `"Trebuchet"`, `"Georgia"`, `"Palatino"` or `"Comic Sans"`.
#' @param color a valid [color string](https://www.imagemagick.org/Magick++/Color.html) such as `"navyblue"` or `"#000080"`
#' @param strokecolor a [color string](https://www.imagemagick.org/Magick++/Color.html) adds a stroke (border around the text)
#' @param gravity string with [gravity](https://www.imagemagick.org/Magick++/Enumerations.html#GravityType) value from [gravity_types].
#' @param size font-size in pixels
#' @examples
#' meme_get("AllTheThings") %>% meme_text_top("GGPLOT ALL THE THINGS!")
#' meme_get("OprahGiveaway") %>% meme_text_bottom("EVERYONE GETS A %>%!!!")
#' @importFrom magick image_annotate geometry_point
#' @inheritDotParams magick::image_annotate location degrees boxcolor
#'
#' @references
#' Jeroen Ooms (2018). magick: Advanced Graphics and Image-Processing in R. R package version 2.0.
#' \url{https://CRAN.R-project.org/package=magick}
#' @export
#' @describeIn meme_text_top Place text in center top of meme
meme_text_top <- function(img, txt, font = "Impact",
                          color = "white", strokecolor = "black",
                          gravity = "North", size = 44, ...) {
  image_annotate(img, txt, font = font,
                 color = color, strokecolor = strokecolor,
                 gravity = gravity, size = size, ...)
}

#' @export
#' @describeIn meme_text_top Place text in bottom center of meme
meme_text_bottom <- function(img, txt, font = "Impact",
                             color = "white", strokecolor = "black",
                             gravity = "South", size = 44, ...) {
  image_annotate(img, txt, font = font,
                 color = color, strokecolor = strokecolor,
                 gravity = gravity, size = size, ...
  )
}

#' Specialized functions for placing text in memes
#'
#' @param newgirl A character string. Text for the left-most person in the `"DistractedBf"` meme.
#' @param guy A character string. Text for the center person in the `"DistractedBf"` meme.
#' @param oldgirl A character string. Text for the right-most person in the `"DistractedBf"` meme.
#' @param font string with font family such as `"sans"`, `"mono"`, `"serif"`, `"Times"`, `"Helvetica"`, `"Trebuchet"`, `"Georgia"`, `"Palatino"` or `"Comic Sans"`.
#' @param color a valid [color string](https://www.imagemagick.org/Magick++/Color.html) such as `"navyblue"` or `"#000080"`
#' @param strokecolor a [color string](https://www.imagemagick.org/Magick++/Color.html) adds a stroke (border around the text)
#' @param size font-size in pixels
#' @importFrom magick image_annotate geometry_point
#' @importFrom stringr str_wrap
#' @inheritDotParams magick::image_annotate location degrees boxcolor
#'
#' @examples
#' meme_get("DistractedBf") %>%
#'   meme_text_distbf("tidyverse", "new R users", "base R")
#' @export
#' @describeIn meme_text_distbf Text function for the distracted boyfriend meme.
meme_text_distbf <- function(img, newgirl, guy, oldgirl, font = "Impact",
                             color = "white", strokecolor = "black", size = 44, ...) {
  image_annotate(img, newgirl,
    font = font,
    color = color, strokecolor = strokecolor,
    gravity = "SouthWest", location = geometry_point(65, 0), size = size, ...
  ) %>%
    image_annotate(guy,
                   font = font,
                   color = color, strokecolor = strokecolor,
      gravity = "Center", location = geometry_point(65, 0), size = size, ...
    ) %>%
    image_annotate(oldgirl,
                   font = font,
                   color = color, strokecolor = strokecolor,
      gravity = "SouthEast", location = geometry_point(10, 30), size = size, ...
    )
}

#' @param img An image of class `magick_image`. See [meme_get()].
#' @param dwayne A character string. Text for the top speech bubble in the `"TheRockDriving"` meme.
#' @param girl A character string. Text for the other speech bubble in the `"TheRockDriving"` meme.
#'
#' @describeIn meme_text_distbf Text function for The Rock driving meme.
#'
#' @examples
#' meme_get("TheRockDriving") %>%
#'   meme_text_rock("What's your favorite thing to do in R?" , "Write for loops.")
#' @export
meme_text_rock <- function(img, dwayne, girl, font = "Impact",
                           color = "black", strokecolor = NULL, size = 16, ...) {
  image_annotate(img, str_wrap(dwayne,20),
                 font = font,
                 color = color, strokecolor = strokecolor,
                 gravity = "North", location = geometry_point(90,20), size = size, ...
  ) %>%
    image_annotate(girl,
                   font = font,
                   color = color, strokecolor = strokecolor,
                   gravity = "Center", location = geometry_point(80,-42), size = size, ...
    )
}

#' @param robin A character string. Text for the Robin's speech bubble in the `"BatmanRobin"` meme.
#' @param batman A character string. Text for the Batman's speech bubble in the `"BatmanRobin"` meme.
#'
#' @describeIn meme_text_distbf Text function for the Batman slaps Robin meme.
#'
#' @examples
#' meme_get("BatmanRobin") %>%
#'   meme_text_batman("Hey, I'm Batman!" , "No, you idiot, I'm Batman!")
#' @export
meme_text_batman <- function(img, robin, batman, font = "Impact",
                             color = "black", strokecolor = NULL, size = 20, ...) {
  image_annotate(img, str_wrap(robin,20),
                 font = font,
                 color = color, strokecolor = strokecolor,
                 gravity = "NorthWest", location = geometry_point(10,0), size = size, ...
  ) %>%
    image_annotate(str_wrap(batman, 20),
                   font = font,
                   color = color, strokecolor = strokecolor,
                   gravity = "NorthEast", location = geometry_point(20,0), size = size, ...
    )
}

#' @param rightpg A character string. Text for the right page in the `"TrumpBillSigning"` meme.
#' @param leftpg A character string. Text for the left page in the `"TrumpBillSigning"` meme.
#'
#' @describeIn meme_text_distbf Text function for the Trump "first order of business" meme.
#'
#' @examples
#' meme_get("TrumpBillSigning") %>%
#'   meme_text_trump("Stuff and nonsense")
#' @export
meme_text_trump <- function(img, rightpg, leftpg = "", font = "Impact",
                            color = "black", strokecolor = NULL, size = 20, ...){
  image_annotate(img, str_wrap(rightpg,12),
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "East", location = geometry_point(60,0), size = size, ...
                ) %>%
    image_annotate(str_wrap(leftpg,12),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "Center", location = geometry_point(0,0), size = size, ...
                   )
}


#' @param isthis A character string. Text for the question in the `"IsThisAPigeon"` meme.
#' @param humanoid A character string. Text for the humanoid in the `"IsThisAPigeon"` meme.
#' @param butterfly A character string. Text for the butterfly in the `"IsThisAPigeon"` meme.
#'
#' @describeIn meme_text_distbf Text function for the Is this a pigeon? meme.
#'
#' @examples
#' meme_get("IsThisAPigeon") %>%
#'   meme_text_pigeon("Is this a pigeon?" , "Me", "Not a pigeon")
#' @export
meme_text_pigeon <- function(img, isthis, humanoid, butterfly, font = "Impact",
                             color = "white", strokecolor = "black", size = 30,
                             ...) {
  image_annotate(img, isthis,
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "South", size = size*1.5, ...
  ) %>%
  image_annotate(str_wrap(humanoid,20),
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "West", location = geometry_point(160,-60), size = size, ...
  ) %>%
    image_annotate(str_wrap(butterfly, 20),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "NorthEast", location = geometry_point(100,100), size = size, ...
    )
}


#' @param sr1 A character string. Text for the first panel in the `"AmericanChopper"` meme.
#' @param jr1 A character string. Text for the second panel in the `"AmericanChopper"` meme.
#' @param sr2 A character string. Text for the third panel in the `"AmericanChopper"` meme.
#' @param jr2 A character string. Text for the fourth panel in the `"AmericanChopper"` meme.
#' @param sr3 A character string. Text for the fifth panel in the `"AmericanChopper"` meme.
#'
#' @describeIn meme_text_distbf Text function for the American Chopper Senior vs. Junior fight meme.
#'
#' @examples
#' meme_get("AmericanChopper") %>%
#'   meme_text_chopper("Stop calling yourself 'just an R user'!",
#'                     "But I've never written a package!",
#'                     "So?? You're still an important part of the R community!",
#'                     "But people who write packages are more important",
#'                     "NO! All members of the R community are valued & deserve respect!")
#' @export
meme_text_chopper <- function(img, sr1, jr1, sr2, jr2, sr3,
                              font = "Impact", color = "white", strokecolor = "black",
                              size = 20, ...){
  image_annotate(img, str_wrap(sr1,35),
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "North", location = geometry_point(5,150), size = size, ...
                 ) %>%
  image_annotate(str_wrap(jr1,35),
                 font = font, color = color, strokecolor = strokecolor,
                gravity = "Center", location = geometry_point(0,-110), size = size, ...
    ) %>%
    image_annotate(str_wrap(sr2,35),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "Center", location = geometry_point(0,50), size = size, ...
                   ) %>%
    image_annotate(str_wrap(jr2,35),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "Center", location = geometry_point(0,250), size = size, ...

    ) %>%
    image_annotate(str_wrap(sr3,35),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "South", location = geometry_point(0,0), size = size, ...
                   )
}


#' @param br1 A character string. Text for the first panel in the `"ExpandingBrain"` meme.
#' @param br2 A character string. Text for the second panel in the `"ExpandingBrain"` meme.
#' @param br3 A character string. Text for the third panel in the `"ExpandingBrain"` meme.
#' @param br4 A character string. Text for the fourth panel in the `"ExpandingBrain"` meme.
#'
#' @describeIn meme_text_distbf Text function for the expanding brain meme.
#'
#' @examples
#' meme_get("ExpandingBrain") %>%
#'   meme_text_brain("text1",
#'                     "text2",
#'                     "text3",
#'                     "text4")
#' @export
meme_text_brain <- function(img, br1, br2, br3, br4,
                            font = "Impact", color = "black", strokecolor = NULL,
                            size = 24, ...){
  image_annotate(img, str_wrap(br1,35),
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "NorthWest", location = geometry_point(10,20), size = size, ...
  ) %>%
    image_annotate(str_wrap(br2,35),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "West", location = geometry_point(10,-130), size = size, ...
    ) %>%
    image_annotate(str_wrap(br3,35), font = "Impact", color = "black", strokecolor = NULL,
                   gravity = "West", location = geometry_point(10,40), size = size, ...
    ) %>%
    image_annotate(str_wrap(br4,35),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "SouthWest", location = geometry_point(10,120), size = size, ...
    )
}


#' @param good A character string. A good idea for the `"OffRamp"` meme.
#' @param dark A character string. A tempting, bad idea for the `"OffRamp"` meme.
#'
#' @describeIn meme_text_distbf Text function for the Off Ramp meme.
#'
#' @examples
#' meme_get("EvilKermit") %>%
#'   meme_text_kermit("Me: Your text here", "Me: Your evil text here")
#'
#' @export
meme_text_kermit <- function(img, good, dark,
                             font = "Impact", color = "black", strokecolor = NULL,
                             size = 20, ...){
  image_annotate(img, str_wrap(good,100),
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "NorthWest", location = geometry_point(10,30), size = size, ...
  ) %>%
    image_annotate(str_wrap(dark,100),
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "NorthWest", location = geometry_point(10,100), size = size, ...
    )
}

#' @param lbtn A character string. The left button in the `"TwoButtonsAnxiety"` meme.
#' @param rbtn A character string. The right button in the `"TwoButtonsAnxiety"` meme.
#'
#' @describeIn  meme_text_distbf Text function for the Two Buttons Anxiety meme.
#'
#' @examples
#' meme_get("TwoButtonsAnxiety") %>%
#'   meme_text_buttons("Should I do this?", "Or this?")
#'
#' @export
meme_text_buttons <- function(img, lbtn, rbtn, font = "Impact",
                              color = "black", strokecolor = NULL,
                              size = 18, ...){
  image_annotate(img, str_wrap(lbtn,15), degrees = -15,
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "NorthWest", location = geometry_point(30,60), size = size, ...
  ) %>%
    image_annotate(str_wrap(rbtn, 15), degrees = -10,
                   font = font, color = color, strokecolor = strokecolor,
                   gravity = "North", location = geometry_point(20,35), size = size, ...
                   )
}

#' @param top A character string. Text for the top panel in the `"HotlineDrake"` meme.
#' @param bot A character string. Text for the bottom panel in the `"HotlineDrake"` meme.
#'
#' @describeIn  meme_text_distbf Text function for the Drake meme.
#'
#' @examples
#' meme_get("HotlineDrake") %>%
#'   meme_text_drake("Handcrafted memes", "Reproducible memes")
#'
#' @export
meme_text_drake <- function(img, top, bot,
                            font = "Impact", color = "black", strokecolor = NULL,
                            size = 30, ...){
  image_annotate(img, str_wrap(top,15),
                 font = font, color = color, strokecolor = strokecolor,
                 gravity = "North", location = geometry_point(125,0), size = size, ...
  ) %>%
    image_annotate(str_wrap(bot,15), font = "Impact", color = "black", strokecolor = NULL,
                   gravity = "Center", location = geometry_point(125,50), size = size, ...
    )
}
